lib/Dialect/FIRRTL/Transforms/GrandCentralTaps.cpp

```diff
@@ -0,0 +1,653 @@
+  //   is unambiguous. Make it relative to the blackbox instance path.
+  // - Look up the port on the black box.
+  // - Create a hierarchical SV name and store this as an assignment to the
+  //   blackbox port.
+  //   - DeletedDataTapKey: skip and don't create a wiring
+  //   - LiteralDataTapKey: just the literal
+  //   - ReferenceDataTapKey: relative path with "." + source name
+  //   - DataTapModuleSignalKey: relative path with "." + internal path
+  // - Generate a body for the blackbox module with the signal mapping
+
+  memPortIdx.clear();
+
+  // Gather some string attributes in the context to simplify working with the
+  // annotations.
+  Strings strings(&getContext());
+
+  // Gather a list of extmodules that have data or mem tap annotations to be
+  // expanded.
+  SmallVector<AnnotatedExtModule, 4> modules;
+  for (auto &op : *circuitOp.getBody()) {
+    auto extModule = dyn_cast<FExtModuleOp>(&op);
+    if (!extModule)
+      continue;
+
+    // Go through the module ports and collect the annotated ones.
+    AnnotatedExtModule result{extModule, {}, {}, {}};
+    result.filteredPortAnnos.reserve(extModule.getNumArguments());
+    for (unsigned argNum = 0, e = extModule.getNumArguments(); argNum < e;
+         ++argNum) {
+      // Go through all annotations on this port and add the data tap key and
+      // mem tap ones to the list.
+      auto annos = AnnotationSet::forPort(extModule, argNum);
+      annos.removeAnnotations([&](Annotation anno) {
+        if (anno.isClass(strings.memTapClass, strings.deletedKeyClass,
+                         strings.literalKeyClass, strings.referenceKeyClass,
+                         strings.internalKeyClass)) {
+          result.portAnnos.push_back({argNum, anno});
+          return true;
+        }
+        return false;
+      });
+      result.filteredPortAnnos.push_back(annos);
+    }
+
+    // If there are data tap annotations on the module, which is likely the
+    // case, create a filtered array of annotations with them removed.
+    AnnotationSet annos(extModule.getOperation());
+    annos.removeAnnotations(
+        [&](Annotation anno) { return anno.isClass(strings.dataTapsClass); });
+    result.filteredModuleAnnos = annos.getArrayAttr();
+
+    if (!result.portAnnos.empty())
+      modules.push_back(std::move(result));
+  }
+
+#ifndef NDEBUG
+  for (auto m : modules) {
+    LLVM_DEBUG(llvm::dbgs() << "Extmodule " << m.extModule.getName() << " has "
+                            << m.portAnnos.size() << " port annotations\n");
+  }
+#endif
+
+  // Fast path if there's nothing to do.
+  if (modules.empty()) {
+    markAllAnalysesPreserved();
+    return;
+  }
+
+  // Build an instance graph.
+  InstanceGraph instanceGraph(circuitOp);
+
+  // Gather the annotated ports and operations throughout the design that we are
+  // supposed to tap in one way or another.
+  tappedArgs.clear();
+  tappedOps.clear();
+  circuitOp.walk([&](Operation *op) {
```

Comment:
I'd recommend splitting this lambda out to its own static function since it is large.